solr/modules/monitor/src/java/org/apache/solr/monitor/search/QueryMatchType.java

```diff
@@ -0,0 +1,45 @@
+/*
+ *
+ *  * Licensed to the Apache Software Foundation (ASF) under one or more
+ *  * contributor license agreements.  See the NOTICE file distributed with
+ *  * this work for additional information regarding copyright ownership.
+ *  * The ASF licenses this file to You under the Apache License, Version 2.0
+ *  * (the "License"); you may not use this file except in compliance with
+ *  * the License.  You may obtain a copy of the License at
+ *  *
+ *  *     http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package org.apache.solr.monitor.search;
+
+public enum QueryMatchType {
```

Comment:
https://github.com/kotman12/solr/pull/3 proposes to defer monitorMatchType support (always use simple matching) i.e. to not have it in the initial integration: that reduces scope and complexity initially code wise etc. and also from users' initial onboarding perspective, perhaps.

_defer_ as opposed to _remove_ i.e. looking at https://lucene.apache.org/core/9_11_0/monitor/org/apache/lucene/monitor/QueryMatch.html documentation there's not only HighlightsMatch but also ExplainingMatch and ScoringMatch.

Reply:
@cpoerschke I responded on the actual PR but thinking I should have replied here instead

Reply:
No worries at all. Perhaps 'main' points, like your https://github.com/kotman12/solr/pull/3#issuecomment-2187092848 comment, could be "here" and 'minor' points, like my file-level https://github.com/kotman12/solr/pull/3/files#r1651440706 remark about the replacement of the `queryMatchType.needsScores` evaluation, could be "there" or something like that.

Picking up from part of the https://github.com/kotman12/solr/pull/3#issuecomment-2187092848 points:

> ... But I think the doc list response is the correct approach because users are already familiar with it. It seamlessly integrates with other solr features ...

Yes, I think so too.

> ... I'm still thinking about how features like highlights or explaining match could fit into this (in a later iteration) ...

I don't yet understand the highlights and explaining matches enough to comment specifically on that, but conceptually document transformers might be a fit: https://solr.apache.org/guide/solr/latest/query-guide/document-transformers.html

Reply:
I can see that working. Either way, it would be best to utilize lucene-monitor's [highlight matches](https://github.com/apache/lucene/blob/3ae59a9809d9239593aa94dcc23f8ce382d59e60/lucene/monitor/src/java/org/apache/lucene/monitor/HighlightsMatch.java#L71-L93) since they come with the character offsets for free. If we pipe that data to a `DocTransformer` (perhaps via `setContext`?), then we can bind it to a doc-level field...  It shouldn't need any index access since the match offsets and document text will suffice. The document text will likely be in the reverse search request itself (and the reverse search _index_ likely won't have the underlying _document_ text anyway unless the same collection is being used). There might be some edge-cases like pre-analyzed document fields which can theoretically be matched in the monitor _without_  text but that is for sophisticated users who probably understand the various limitations already.

As an aside, do you know why the solr highlighters [don't seem to use a DocTransformer](https://github.com/apache/solr/blob/709a1ee27df23b419d09fe8f67c3276409131a4a/solr/core/src/java/org/apache/solr/handler/component/HighlightComponent.java#L159-L167) (not directly at least)? Prior to you linking this feature I assumed reverse search highlights would adhere to a similar API to existing highlights with that separate top-level `highlighting` field. But I can also see how this is a bit awkward and perhaps "vestigial" ... also there is an interesting 16-year old `// TODO` in that section, perhaps referring to the same thing 😃

Reply:
> As an aside, do you know why the solr highlighters [don't seem to use a DocTransformer](https://github.com/apache/solr/blob/709a1ee27df23b419d09fe8f67c3276409131a4a/solr/core/src/java/org/apache/solr/handler/component/HighlightComponent.java#L159-L167) (not directly at least)?

Good question, I don't know but perhaps @dsmiley might.

> ... perhaps "vestigial" ... 

I re-learnt an interesting word there :-)

> also there is an interesting 16-year old `// TODO` in that section, perhaps referring to the same thing 😃

Once there's an existing way I guess it's a little tricky to move away from that, with backwards compatibility (adding a new alternative way, deprecating-but-retaining-the-existing way, then removing the existing way) but maybe there's also other reasons/benefits of using a top-level highlighting fields, I haven't thought about it much as yet.

Reply:
https://issues.apache.org/jira/browse/SOLR-3479 PR welcome :-).  DocTransformer came out after SearchComponent based Highlighter existed.
It would probably be not too hard to do one using the UnifiedHighlighter.  It would also enable massive streaming response use-cases that can't easily be done today.

Reply:
>> ... But I think the doc list response is the correct approach because users are already familiar with it. It seamlessly integrates with other solr features ...

> Yes, I think so too.

@cpoerschke the tests reminded me that another motivation for the top-level `monitor` response object is to handle the "batch" document response, i.e.:

```json
{
    "monitorDocuments": {
              "0": { "queries": [1]}
              "1": {"queries":  [1, 2, 5]}
    }
}
```
where a user matches many documents and expects the query matches to be bucketed per requested document. This is currently part of the original monitor API although it's arguable whether it is in scope for this initial solr integration.

The reason I bring it up is because your [PR](https://github.com/kotman12/solr/pull/3) inspired me to try to remove the query match type altogether and that is one thing I ran into. I especially wanted to get rid of the noisy custom codec and merging logic. I was secretly hoping there would be a minimal way to add back the highlighting in a more integrated form (perhaps via some clever `ReturnFields`/`DocTransformer` rigging in the `ReverseSearchComponent`) after the de-noising but it might be challenging to do it minimally.

Reply:
> ... matches many documents and expects the query matches to be bucketed per requested document. This is currently part of the original monitor API although it's arguable whether it is in scope for this initial solr integration. ...

Hmm, yes, that's an interesting question. I think my thoughts on this would be something along these lines:
* Is matching multiple documents at the same time a common use case?
* If it's not in scope for the initial integration then it could be added later potentially.
* Single document makes for a simpler request and response interface from the user's perspective and from the code perspective too e.g. no need to check if there's zero documents in the batch or duplicates in the batch.
* Multiple documents, or the potential for multiple documents, reduces familiarity with i.e. similarity to existing functionality i.e. it would be somewhat akin to supporting more than one `q` parameter in a search. And if (say) _two_ documents are supplied, does it (from the user's point of view) mean to match against _either of them_ or does it mean to match against _both of them_?
* Nested documents -- I don't know much about them yet actually -- might they provide a way (beyond the initial integration here) to support multiple-documents logic via a single document?
  * https://solr.apache.org/guide/solr/latest/indexing-guide/indexing-nested-documents.html